tests/debug-tests/test_endpoint_error_callback.py

```diff
@@ -0,0 +1,179 @@
+from distributed.protocol import to_serialize
+
+import cloudpickle
+import pytest
+import ucp
+from ucp.utils import get_ucxpy_logger
+from utils import get_cuda_devices, get_num_gpus, recv, send
+
+cupy = pytest.importorskip("cupy")
+
+UCX_TLS = os.environ.get("UCX_TLS")
+rc_enabled = UCX_TLS is not None and "rc" in UCX_TLS
+
+
+def get_log_queue_handler():
+    handler = logging.StreamHandler()
+
+    log_queue = queue.Queue(-1)  # no limit on size
+    queue_handler = logging.handlers.QueueHandler(log_queue)
+    listener = logging.handlers.QueueListener(log_queue, handler)
+
+    ucxpy_logger = get_ucxpy_logger()
+    ucxpy_logger.addHandler(queue_handler)
+
+    formatter = logging.Formatter("%(message)s")
+    handler.setFormatter(formatter)
+
+    return log_queue, listener
+
+
+async def get_ep(name, port):
+    addr = ucp.get_address()
+    ep = await ucp.create_endpoint(addr, port)
+    return ep
+
+
+def client(port, func):
+    # wait for server to come up
+    # receive object
+    # process suicides
+
+    ucp.init()
+
+    # must create context before importing
+    # cudf/cupy/etc
+
+    async def read():
+        await asyncio.sleep(1)
+        ep = await get_ep("client", port)
+        msg = None
+        import cupy
+
+        cupy.cuda.set_allocator(None)
+
+        frames, msg = recv(ep)
+
+        # Client process suicides to force an "Endpoint timeout"
+        # on the server
+        os.kill(os.getpid(), signal.SIGKILL)
+
+    asyncio.get_event_loop().run_until_complete(read())
+
+
+def server(port, func):
+    # create listener receiver
+    # add queue logger
+    # write cudf object
+    # terminates ep/listener
+    # checks that "Endpoint timeout" was logged
+    ucp.init()
+
+    log_queue, log_listener = get_log_queue_handler()
+    log_listener.start()
+
+    async def f(listener_port):
+        # coroutine shows up when the client asks
+        # to connect
+        async def write(ep):
+            import cupy
+
+            cupy.cuda.set_allocator(None)
+
+            print("CREATING CUDA OBJECT IN SERVER...")
+            cuda_obj_generator = cloudpickle.loads(func)
+            cuda_obj = cuda_obj_generator()
+            msg = {"data": to_serialize(cuda_obj)}
+            frames = await to_frames(msg, serializers=("cuda", "dask", "pickle"))
+
+            # Send meta data
+            try:
+                await send(ep, frames)
+            except Exception:
+                # Avoids process hanging on "Endpoint timeout"
+                pass
+
+            print("Shutting Down Server...")
+            await ep.close()
+            lf.close()
+
+        lf = ucp.create_listener(write, port=listener_port)
+        try:
+            while not lf.closed():
+                await asyncio.sleep(0.1)
+        except ucp.UCXCloseError:
+            pass
+
+    log_listener.stop()
+
+    asyncio.get_event_loop().run_until_complete(f(port))
+
+    # Check log for the expected "Endpoint timeout" and exits with
```

Comment:
Thank you for adding this test Peter!  It might be worth mentioning that without `UCP_ERR_HANDLING_MODE_PEER` we would not see an endpoint timeout and instead would see fatal error.

Reply:
I honestly don't see the point in writing that. The test is specifically testing the endpoint error callback, seems redundant to me.

Reply:
That's true.  I suppose the name of the test `test_endpoint_error_callback` is a large enough clue.

Reply:
On the other hand, I realized we didn't test without the endpoint error callback, which we can do and make it clear. Let me try to add that.

Reply:
Ok, I added the test, hopefully it makes more sense now.